easypost/services/base_service.py

```diff
@@ -0,0 +1,101 @@
+import re
+from typing import (
+    Any,
+    Dict,
+    List,
+    Optional,
+)
+
+from easypost.easypost_object import convert_to_easypost_object
+from easypost.error import Error
+from easypost.requestor import (
+    RequestMethod,
+    Requestor,
+)
+
+
+class BaseService:
+    """The base service that all other services inherit."""
+
+    def __init__(self, client):
+        self.client = client
+
+    def snakecase_name(self, class_name) -> str:
+        """Return the class name as snake_case."""
+        return re.sub(r"(?<!^)(?=[A-Z])", "_", class_name).lower()
+
+    def class_url(self, class_name) -> str:
+        """Generate a URL based on class name."""
+        if class_name[-1:] in ("s", "h"):
+            return f"/{class_name}es"
+        else:
+            return f"/{class_name}s"
+
+    def instance_url(self, class_name, id) -> str:
+        """Generate an instance URL based on the ID of the object."""
+        if not id:
+            raise Error(f"{class_name} instance has invalid ID: {id}")
+
+        class_url = self.class_url(class_name)
+
+        return f"{class_url}/{id}"
+
+    def create_resource(self, class_name: str, **params) -> Any:
+        """Create an EasyPost object."""
+        url = self.class_url(class_name)
+        wrapped_params = {self.snakecase_name(class_name): params}
+
+        # TODO: Don't instantiate the Requestor class
+        response, api_key = Requestor().request(method=RequestMethod.POST, url=url, params=wrapped_params)
+
+        # TODO: Get rid of the api_key
+        return convert_to_easypost_object(response=response, api_key=api_key)
+
+    def all_resources(self, class_name: str, **params) -> List[Any]:
+        """Retrieve a list of records from the EasyPost API."""
+        url = self.class_url(class_name)
+        # TODO: Don't instantiate the Requestor class
+        response, api_key = Requestor().request(method=RequestMethod.GET, url=url, params=params)
+
+        # TODO: Get rid of the api_key
+        return convert_to_easypost_object(response=response, api_key=api_key)
+
+    def retrieve_resource(self, class_name: str, id: str) -> Any:
+        """Retrieve an object from the EasyPost API."""
+        url = self.instance_url(class_name, id)
+        # TODO: Don't instantiate the Requestor class
+        response, api_key = Requestor().request(method=RequestMethod.GET, url=url)
```

Comment:
Issue: There is a trade-off in making these general functions (create, retrieve, all) in the parent class, this will allow users to use these functions on a service that doesn't support it.

Reply:
I can make them "internal" by prepending an underscore? There isn't really a great way in Python to make something private and not accessible to the users.

Reply:
Made them all underscored, let me know if that's not a good idea. The problem with the class inheritance we had before was it was a bit magical and didn't allow for us to be explicit in each function call which we now need to be because of the client object (need to pass in each model string, etc)

Reply:
If we don't want users to use these internal functions or just show up in their IDE auto-completion. We should be explicit for each service (that's what I am doing in Ruby) but more code for the lib. There is a trade-off decision we need to think about, would like more feedback from other people on this.

Reply:
I agree with putting them in the parent class (to keep the code DRY) and making them "internal" by marking with the leading underscore. Python as a language isn't good with access levels, but we can do our best to work within the conventions of the language, under the assumption that our end-users are respecting those. If an end-user chooses to access an underscored method and odd behavior happens as a result, that's on them at that point.